resources/js/components/textarea-autosize.tsx

```diff
@@ -0,0 +1,79 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the GNU Affero General Public License v3.0.
+// See the LICENCE file in the repository root for full licence text.
+
+import autosize from 'autosize';
+import React from 'react';
+
+interface Props extends React.TextareaHTMLAttributes<HTMLTextAreaElement> {
+  async: boolean;
+  innerRef?: React.RefObject<HTMLTextAreaElement>;
+  maxRows?: number;
+}
+
+interface State {
+  lineHeight?: number;
+}
+
+export default class TextareaAutosize extends React.Component<Props, State> {
+  static readonly defaultProps = {
+    async: false,
+    rows: 1,
+  };
+
+  private ref = this.props.innerRef ?? React.createRef<HTMLTextAreaElement>();
+
+  private get maxHeight() {
+    return this.props.maxRows != null && this.state.lineHeight != null
+      ? Math.ceil(this.state.lineHeight * (this.props.maxRows + 1)) // additional row to fit maxRows + slight leeway for scrollbar
+      : null;
+  }
+
+  constructor(props: Props) {
+    super(props);
+    this.state = {};
+  }
+
+  componentDidMount() {
+    if (this.ref.current == null) return;
+
+    if (this.props.maxRows != null || this.props.async) {
+      window.setTimeout(() => {
+        if (this.ref.current != null) {
+          if (this.props.maxRows != null) {
+            // getting line-height should be delayed until after turbolinks navigation, otherwise it returns NaN.
+            const lineHeight = parseFloat(window.getComputedStyle(this.ref.current).getPropertyValue('line-height'));
+            this.setState({ lineHeight });
+          }
+
+          autosize(this.ref.current);
+        }
+      }, 0);
+    } else {
+      autosize(this.ref.current);
+    }
+  }
+
+  componentDidUpdate() {
+    if (this.ref.current == null) return;
+    autosize.update(this.ref.current);
+  }
+
+  componentWillUnmount() {
+    if (this.ref.current == null) return;
+    autosize.destroy(this.ref.current);
+  }
+
+  render() {
+    const { async, innerRef, maxRows, style, ...otherProps } = this.props;
+
+    const maxHeight = this.maxHeight;
+
+    return (
+      <textarea
```

Comment:
not sure exactly where but the styling seems to be slightly wrong

old
![](https://s.myconan.tk/2023/09/firefox_2023-09-11_22-02-50.png)

this pr
![](https://s.myconan.tk/2023/09/firefox_2023-09-11_22-04-32.png?a)

note the wrong background colour for the padding

<details>
<summary>it's slightly more <em>interesting</em> in chrome</summary>

![](https://s.myconan.tk/2023/09/chrome_2023-09-11_22-06-26.png)

</details>

Reply:
The old round uses `Math.round` before multiplying the rows, this one applies `Math.ciel` after multiplying; current `line-height` is `17.5px`

Reply:
white box appears to be a change in `autosize` between 4 and the current version where previously it applied `overflow: hidden scroll` but not it doesn't 🤔